File: app/src/main/java/com/radiozkousky/data/ProgressStore.kt
package com.radiozkousky.data

import android.content.Context
import android.content.SharedPreferences
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

data class QuestionProgress(
    val questionId: Int = 0,
    var correctCount: Int = 0,
    var incorrectCount: Int = 0,
    var lastReviewedAt: Long = 0L,
    var nextReviewAt: Long = 0L,
    var easeFactor: Double = 2.5,
    var interval: Int = 0,
    var repetition: Int = 0
)

class ProgressStore(context: Context) {

    private val prefs: SharedPreferences =
        context.getSharedPreferences("radio_zkousky_progress", Context.MODE_PRIVATE)
    private val gson = Gson()

    private var progressMap: MutableMap<Int, QuestionProgress> = loadAll()

    private fun loadAll(): MutableMap<Int, QuestionProgress> {
        val json = prefs.getString("progress_data", null) ?: return mutableMapOf()
        val type = object : TypeToken<MutableMap<Int, QuestionProgress>>() {}.type
        return try {
            gson.fromJson(json, type) ?: mutableMapOf()
        } catch (e: Exception) {
            mutableMapOf()
        }
    }

    private fun saveAll() {
        val json = gson.toJson(progressMap)
        prefs.edit().putString("progress_data", json).apply()
    }

    fun getProgress(questionId: Int): QuestionProgress {
        return progressMap.getOrPut(questionId) { QuestionProgress(questionId = questionId) }
    }

    fun recordAnswer(questionId: Int, correct: Boolean) {
        val progress = getProgress(questionId)
        if (correct) {
            progress.correctCount++
        } else {
            progress.incorrectCount++
        }
        progress.lastReviewedAt = System.currentTimeMillis()
        SpacedRepetition.updateProgress(progress, correct)
        progressMap[questionId] = progress
        saveAll()
    }

    fun getAllProgress(): Map<Int, QuestionProgress> = progressMap.toMap()

    fun getCorrectCount(questionId: Int): Int = getProgress(questionId).correctCount

    fun getTotalAnswered(): Int = progressMap.values.count { it.correctCount + it.incorrectCount > 0 }

    fun getTotalCorrect(): Int = progressMap.values.sumOf { it.correctCount }

    fun getTotalIncorrect(): Int = progressMap.values.sumOf { it.incorrectCount }

    fun getMasteredCount(): Int = progressMap.values.count { it.repetition >= 3 }

    fun getMasteredCountByCategory(category: Category): Int {
        val questionIds = QuestionBank.getByCategory(category).map { it.id }.toSet()
        return progressMap.values.count { it.questionId in questionIds && it.repetition >= 3 }
    }

    fun getQuestionsForReview(category: Category? = null): List<Question> {
        val now = System.currentTimeMillis()
        val questions = if (category != null) {
            QuestionBank.getByCategory(category)
        } else {
            QuestionBank.allQuestions
        }

        val dueQuestions = questions.filter { q ->
            val progress = progressMap[q.id]
            progress == null || progress.nextReviewAt <= now
        }

        return if (dueQuestions.isEmpty()) {
            questions.shuffled().take(10)
        } else {
            dueQuestions.sortedBy { q ->
                progressMap[q.id]?.nextReviewAt ?: 0L
            }
        }
    }

    fun resetAll() {
        progressMap.clear()
        prefs.edit().clear().apply()
    }
}

File: web/src/lib/progressStore.ts
import { Category, ExamType, Question, QuestionProgress } from '../data/types'
import { getAllQuestions, getByCategory } from '../data/questionBank'
import { createProgress, updateProgress } from './spacedRepetition'

const STORAGE_PREFIX = 'radio_zkousky_progress_'

function getStorageKey(examType: ExamType): string {
  return `${STORAGE_PREFIX}${examType}`
}

function loadAll(examType: ExamType): Record<number, QuestionProgress> {
  try {
    const json = localStorage.getItem(getStorageKey(examType))
    if (!json) return {}
    return JSON.parse(json)
  } catch {
    return {}
  }
}

function saveAll(examType: ExamType, data: Record<number, QuestionProgress>): void {
  localStorage.setItem(getStorageKey(examType), JSON.stringify(data))
}

export function getProgress(examType: ExamType, questionId: number): QuestionProgress {
  const all = loadAll(examType)
  return all[questionId] ?? createProgress(questionId)
}

export function recordAnswer(examType: ExamType, questionId: number, correct: boolean): void {
  const all = loadAll(examType)
  const current = all[questionId] ?? createProgress(questionId)
  all[questionId] = updateProgress(current, correct)
  saveAll(examType, all)
}

export function getAllProgress(examType: ExamType): Record<number, QuestionProgress> {
  return loadAll(examType)
}

export function getTotalAnswered(examType: ExamType): number {
  const all = loadAll(examType)
  return Object.values(all).filter((p) => p.correctCount + p.incorrectCount > 0).length
}

export function getTotalCorrect(examType: ExamType): number {
  const all = loadAll(examType)
  return Object.values(all).reduce((sum, p) => sum + p.correctCount, 0)
}

export function getTotalIncorrect(examType: ExamType): number {
  const all = loadAll(examType)
  return Object.values(all).reduce((sum, p) => sum + p.incorrectCount, 0)
}

export function getMasteredCount(examType: ExamType): number {
  const all = loadAll(examType)
  return Object.values(all).filter((p) => p.repetition >= 3).length
}

export function getMasteredCountByCategory(examType: ExamType, category: Category): number {
  const all = loadAll(examType)
  const questionIds = new Set(getByCategory(examType, category).map((q) => q.id))
  return Object.values(all).filter((p) => questionIds.has(p.questionId) && p.repetition >= 3).length
}

export function getQuestionsForReview(examType: ExamType, category?: Category): Question[] {
  const now = Date.now()
  const all = loadAll(examType)
  const questions = category ? getByCategory(examType, category) : getAllQuestions(examType)

  const due = questions.filter((q) => {
    const p = all[q.id]
    return !p || p.nextReviewAt <= now
  })

  if (due.length === 0) {
    return [...questions].sort(() => Math.random() - 0.5).slice(0, 10)
  }

  return due.sort((a, b) => {
    const pa = all[a.id]?.nextReviewAt ?? 0
    const pb = all[b.id]?.nextReviewAt ?? 0
    return pa - pb
  })
}

export function resetAll(examType: ExamType): void {
  localStorage.removeItem(getStorageKey(examType))
}
